### ISSUE TITLE
indicate locked cache in OKAPI result

### ISSUE THREAD

**Issue by Bananeweizen** [opencaching/okapi#350]
At the time of writing, the cache OC1275A is locked. When trying to access it via OKAPI, the following error is returned:

{"error":{"developer_message":"Parameter 'cache_code' has invalid value: This cache does not exist.","reason_stack":["bad_request","invalid_parameter"],"status":400,"parameter":"cache_code","whats_wrong_about_it":"This cache does not exist.","more_info":"http://www.opencaching.de/okapi/introduction.html#errors"}}

It would be nice, if the OKAPI could actually indicate the real cause instead (cache is locked). It's confusing, if a client program opens an official cache link, and afterwards displays a "cache does not exist" error.

original discussion: http://forum.opencaching.de/index.php?topic=4349.msg55694#msg55694

**Comment by wrygiel** [comment]
Are you displaying the "whats_wrong_about_it" field to the end user? It was supposed to be an informative field for the eyes of the developers only (and it is always in English).

Currently OKAPI simply cannot access locked caches. It was a security-related design decision. Of course, we could change that, but this would require some work, and the benefit is minimal.

Perhaps you'd be satisfied with a changed description of the "whats_wrong_about_it" field? You are right that the current description is not accurate. "This cache does not exist or is not accessible to OKAPI" would be better.

**Comment by following5** [comment]
> Currently OKAPI simply cannot access locked caches (they "look" as if they didn't exist). It was a security-related design decision

There are to kinds of locked caches: 
* locked by the OC team because of some rules violation (only visible for admins and the owner);
* locked by the owner or the OC team to prevent further logs for this cache (visible for anyone, but hidden on the map).

The second kind is available only at OCDE. When adapting OKAPI for OCDE, we opted for a simple implementation and decided to treat all locked caches the same. So it may happen e.g. that a valid URL points to a locked, visible cache at Opencaching.de, but services/caches/search/by_urls will give a negative result.

**Comment by wrygiel** [comment]
"visible for anyone, but hidden on the map" - I didn't know about that, and I think it is a problem.

If it works this way on OCDE, then I think that such caches should be accessible via geocaches method, while being excluded from all search results. It seems simple enough, but you would need to add a new status for such caches. The "geocache" method states that "Valid cache status codes *currently* include", so I guess we could add a new status without breaking backward compatibility.

**Comment by following5** [comment]
See #123: You had concerns regarding backward compatibility, therefore we did not simply add a status "locked". What has changed since then is that the locked/visible caches have been removed from the map, and may be removed from search as well.

I think that the concerns are still valid: OKAPI docs give no advice on how to handle unknown status, so existing applications may fall back e.g. to "Available". But I also agree that a status "locked" would be nice, and since these caches are hidden from search results, they will be rarely requsted via OKAPI, and we may accept some unexpeced client behaviour.

**Comment by wrygiel** [comment]
Yes, I guess you're right - some apps could break... Perhaps we should implement a safer option: Add the "status2" field and deprecate the "status" field (as we did with the size/size2 fields some time before). The backward-compatible "status" field would respond "Archived" whereas "status2" would go with "Locked".

**Comment by following5** [comment]
I still don't like the idea of locked caches to appear as "archived" in all existing OKAPI clients. Therefore I have another proposal:

* We add an option "included_locked_caches" with default "false" to all methods that retrieve cache data.
* If set to "true", caches with status "locked/visible" will be included in the results and indicated by the new status "Locked". This currently applies only to OCDE.
* As GPX etc. do not know the status "locked", I think we have no choice than translating locked to archived there. That's what OC.de does.
* Tile maps will exclude locked caches.
* Replication will include locked caches.
* caches/geocache and caches/search/by_urls will return a specific error message if a cache cannot be retrieved because it is locked, as requested by Bananenweizen. E.g. "The geocache OCxxxx is locked cannot be retrieved because your application does not support locked caches."

So if an OKAPI client is aware of locked/visible caches, it can deliberately request them.

There is one glitch: Replication data will include the formerly unknown status "Locked". Applications using this feature may break, but I think that is acceptable because it is a rarely used feature, not included in any common geocaching apps.

**Comment by following5** [comment]
> We add an option "include_locked_caches" with default "false" to all methods that retrieve cache data.

This is not needed for caches/search/all, caches/search/bbox and caches/search/nearest, caches/search/save and caches/shortcuts/search_and_retrieve, which already have a status selection option. So the new option would be added to caches/geocache, caches/geocaches, caches/formatters/* and caches/search/by_urls.

**Comment by wrygiel** [comment]
> This is not needed for caches/search/all, caches/search/bbox, caches/search/nearest, caches/search/save and caches/shortcuts/search_and_retrieve, which already have a status selection option.

But the `status` parameter of these method does not accept a "Locked" value. If you mean to add it, then we'd need to "officially" add this status to the list of allowed statuses.

**Comment by wrygiel** [comment]
I think this requires a new `status2` field for caches, and similar `status2` parameters in all seatch methods. All other options have flaws.

**Comment by wrygiel** [comment]
If we decide that `Locked` caches are to be searchable, then:

> caches/geocache and caches/search/by_urls will return a specific error message if a cache cannot be retrieved because it is locked, as requested by Bananenweizen.

These method should simple return a `Locked` cache. I see no need to add a special parameter.

> We add an option "include_locked_caches" with default "false" to all methods that retrieve cache data. If set to "true", caches with status "locked/visible" will be included in the results and indicated by the new status "Locked". This currently applies only to OCDE.

Clients would need to include `Locked` in their `status2` parameter in order to search for locked caches. (It will be impossible to search for locked caches with the "old" `status` parameter.)

> As GPX etc. do not know the status "locked", I think we have no choice than translating locked to archived there. That's what OC.de does.

Since GPX formatter uses the `status` field (not the new `status2` field), nothing will be needed to be changed here.

> Tile maps will exclude locked caches.

Tile maps should include all the caches the user wants them to include. I we are allowing for `Locked` caches to be searched, they should also be searchable using tile maps.

> Replication will include locked caches.

Ok, but the `status` parameter for them should be `Archived`.

The `status` fields and parameters may be marked as deprecated, but they must keep working.

**Comment by following5** [comment]
It looks like it did not point out the problem with status/status2 clearly enough. I will try again:

If we simply enable searching for locked caches and go for the status/status2 solution, then all existing OKAPI clients will immediatly present locked caches as archived caches, which is *wrong*. It is as wrong as presenting archived caches as "temporarily not available". And they will present locked caches e.g. on maps, which is deprecated.

My propsal - which of course is more expensive - solves this problem. It eliminates the need for `status2`, allows adding "Locked" to all `status` options and fields without breaking anything and strongly encourages developers to present "Locked" as "Locked" or not at all, but not wrongly as "Archived". The only flaw I am aware of is the OCE replication thing, which I consider acceptable.

**Comment by following5** [comment]
Oops, forget what I just wrote. Now I undestand what you mean. But, there is still a problem with your propsal: When explicitly requesting geocaches via geocache(s) or search/by_url methods, locked caches will wrongly end up as archived in the presentation of existing OKAPI clients.

Afaik the reason for opening this issue was related to search/by_url: An URL from the OC website, e.g. from the list of an owner's caches, is passed to OKAPI search/by_url, and fails because the cache is locked. Having it return the cache with the  `status` = "Locked" instead to my opinion would also be a failure.

**Comment by following5** [comment]
>> Replication will include locked caches.
> Ok, but the status parameter for them should be Archived.

That would mean that replication would alter the status of locked caches for all existing applications, which I consider utterly wrong.

**Comment by wrygiel** [comment]
> But, there is still a problem with your propsal: When explicitly requesting geocaches via geocache(s) or search/by_url methods, locked caches will wrongly end up as archived in the presentation of existing OKAPI clients.

No, I think it's okay. "by_url" method will return `cache_codes`, and the the client will call `geocache` or `geocaches` with these `cache_codes`, but it will not use the `status2` field - OKAPI may detect that and add the "LOCKED: " prefix to cache names.

**Comment by wrygiel** [comment]
> That would mean that replication would alter the status of locked caches for all existing applications, which I consider utterly wrong.

Well... as you said yourself, not many applications use this data source. You could email their authors and let them know about this issue.

**Comment by wrygiel** [comment]
> OKAPI may detect that and add the "LOCKED: " prefix to cache names.

It may also add an extra paragraph in cache description, e.g.

Notice: The status of this geocached is "Locked". Your application probably does not recognize "Locked" caches and displays them as "Archived".

**Comment by following5** [comment]
>> OKAPI may detect that and add the "LOCKED: " prefix to cache names.

>It may also add an extra paragraph in cache description, e.g.

Okay, that's a hack but a smart one.

Regarding replication, there currently are just two active clients at OC.de, both from Poland. This should be manageable.

**Comment by following5** [comment]
The `status` field is a default field for the geocaches method, and we cannot include `status2` by default. A deprecated field would be included by default, the replacement field not. This is *so* ugly.

Why do you think my proposal with a `included_locked_caches` option is flawed? I see less flaws there than with the status/status2 solution:

* It avoids the hack of tampering with cache names and descriptions.
* It keeps the default fields for the geocache(s) methods clean.
* It makes it less likely that a locked cache will erroneusly be treated by application logic as "archived".
* It emphasizes that locked caches should generally be ignored and only requested/presented if they are needed for reference.

**Comment by following5** [comment]
Ok, we might make the geocache(s) default fields dependent on what status was requested: If status2 is requsted the default changes from `status` to `status2`. Another hack. Do you like that?

**Comment by wrygiel** [comment]
> This is so ugly.

I don't think it's that ugly. It's natural to deprecate things. BTW,
probably none of the applications use defaults anyway.

> Why do you think my proposal with a included_locked_caches option is flawed?

Well... the thing I like the least is that it's inconsistent: If we
have a field and parameter by the same name, then I think they must
behave in the same way (take the same values), *regardless of the
method called*. I've been sticking to this rule for all search
parameters so far. Your solution would break this rule for the
"status" parameter, which is one of the most important ones.

Such inconsistencies may lead to weird mistakes and errors later on. I
really would like to avoid them.

> It avoids the hack of tampering with cache names and descriptions.

I would call it "graceful degradation". :)

> It keeps the default fields for the geocache(s) methods clean.

But it requires newer clients to add a new parameter to *all*
geocache(s) method calls. I think it's better to keep the calls clean.

> It makes it less likely that a locked cache will erroneusly be treated by
> application logic as "archived".

It also does not allow for existing applications to access locked
caches, even they try to access them by specific cache_code. *All*
applications would need to be updated in order to access locked
caches. We should never force developers to update their applications
in order to avoid errors. This is what backward-compatibility is
about.

> It emphasizes that locked caches should generally be ignored and only
> requested/presented if they are needed for reference.

We can emphasize that in the description of the "Locked" status2
value/parameter.



-- 
Hard work has a future payoff. Laziness pays off now.
http://rygielski.net/

**Comment by wrygiel** [comment]
>
> Ok, we might make the geocache(s) default fields dependent on what status
> was requested: If status2 is requsted the default changes from status to
> status2. Another hack. Do you like that?
>
I don't understand what you mean - if the developer requests `status2`
field, then he already doesn't use the default fields.


-- 
Like I said before, I never repeat myself.
http://rygielski.net/

**Comment by following5** [comment]
>  If we have a field and parameter by the same name, then I think they must behave in the same way (take the same values), *regardless of the method called* ... Your solution would break this rule for the "status" parameter, which is one of the most important ones.

Uhmmm...okay. An inconsistency would arise between searching for caches and retrieving them.

> I don't understand what you mean - if the developer requests `status2` field, then he already doesn't use the default fields.

Ok, this problem only applies to search_and_retrieve. When searching for Locked caches, their status will by default be returned as `status`=Archived.

> It also does not allow for existing applications to access locked caches, even if they access them by specific cache_code. *All* applications would need to be updated in order to access locked caches. We should never force developers to update their applications in order to avoid errors. 

My point of view is that the current OKAPI does not define what a locked cache is, and current applications do not know about locked caches and cannot handle them properly (e.g. disable logging and hiding them on maps). Therefore, existing applications should not be able to retrieve them. I think the only error is the error message "cache does not exist"; ist should be something like "this cache cannot be retrieved because it is locked and your application cannot handle locked caches".

This is another thing as size/size2 etc. It does not affect the client's application logic if a nano cache is labeled as micro.

**Comment by following5** [comment]
Please see also the old discussion in #123. I wrote on 22 August 2012, after a discussion in the OC.de team:

> 1. To avoid implementation problems, locked caches - type 6 *and* 7 - should 
gernerally not be accessible by OKAPI.
> 
> 2. We may decide later to add an option like "include_locked=true" for clients 
which want to download locked (status 6) caches; this would include a new cache 
status "Locked" for OKAPI. 

You answered: 
> So, to clear things up:
> You want only caches with status 1..3 accessible via OKAPI. As to all
other caches - OKAPI behaves as if they don't exist at all.

I replied:
> Yes. For types 4/5/7 this is exactly what the OC.de software does when 
referencing caches by waypoints, and for type 6 it is acceptable.

Obviously it was an error to treat them as if they don't exist at all; it lead to a confusing error message which triggered this issue.

**Comment by wrygiel** [comment]
> Ok, this problem only applies to search_and_retrieve. When searching for Locked caches, their status will by default be returned as status=Archived.

He will be forced to use the `status2` parameter when searching, so he will also be aware that `status` parameter has been deprecated. Based on that, I believe all developers should (intuitively) suspect that the `status` field has been deprecated too (even without looking at the `geocache` method specs).

This is the way we've done it previously with other outdated fields and parameters, and I strongly feel we should stick to this.

> This is another thing as size/size2 etc. It does not affect the client's application logic if a nano cache is labeled as micro. And the Locked status is an old and essential feature of OCDE, not some new invention.
> (...)
> Obviously it was an error to treat them as if they don't exist at all; it lead to a confusing error message which triggered this issue.

Sorry about that. :( I didn't know about them back then (none of the OCPL developers was aware of them too).

**Comment by following5** [comment]
>>   If we have a field and parameter by the same name, then I think they must behave in the same way (take the same values), regardless of the method called ... Your solution would break this rule for the "status" parameter, which is one of the most important ones.

> Uhmmm...okay. An inconsistency would arise between searching for caches and retrieving them. I will spend some more thoughts on this.

The status/status2 solution will also create an inconsistency: `status`=Archived will mean "only archived caches" when searching but "cache is archvied or locked" when retrieving.

**Comment by wrygiel** [comment]
That's true! Still, having inconsistencies in deprecated features is better than having them in non-deprecated ones.

**Comment by wrygiel** [comment]
BTW, if you feel uncomfortalbe with this one, I can implement this change.

**Comment by wrygiel** [comment]
> status=Archived will mean "only archived caches" when searching

BTW, are you sure that you don't want Locked caches to appear in search results in older applications? They still would display them with "LOCKED:" prefixes.

**Comment by following5** [comment]
*repeat*  "My point of view is that current applications do not know about locked caches and cannot handle them properly. Therefore, existing applications should not be able to retrieve them."

Yes, I prefer to force developers to update their applications if they want to retrieve locked caches. This forces them to think about what a locked cache is and how to handle it or if it should be ignored, and it is consistent with the current OKAPI implementation. The only objective of locked/visible cache listings is to keep histories and statistics consistent; many clients wont' need them.

The "LOCKED:" prefix (I would prefer a "[locked]" suffix because it does not disturb alphabetical sorting and draws less attention to caches which should usually be ignored) only hafly solves the `status` problem. Also, it can break applications that compare cache names from different sources, e.g. to heuristically identify crosslistings.

On the other hand, I understand your concerns regarding orthogonality of the API. So far it looks to me as if we cannot introduce locked caches without breaking several things. (That's the reason why they are not available yet. ;-)

**Comment by following5** [comment]
This is very hard to decide for me. All alternatives so far have drawbacks, including the one to stay with the status quo and just fix the error message.

I have now reevaluated everything under the two aspects of (a) long-term sustainability and (b) strategic goals of Opencaching.de, which currently is the only OC site affected by this issue. And the status/status2 proposal turned out to probably have an overall advantage in both fields. 

So I am ready to go for status/status2, but still think that we should avoid delivering locked caches to existing applications if possible. And encourage the developers to switch to `status2` as strongly as possible.

**Comment by following5** [comment]
There are some more cache status that are visible to certain users:

* hidden by approvers to check (visible for owners and [some?] admins)
* not published yet (visible for owners)
* "blocked by rr" (OCPL) / "locked/invisible" (OCDE) - visible for owners and [some?] admins

Should we add them to `status2`, to avoid a `status3` later? Fully implementing this makes things more complicated.

**Comment by wrygiel** [comment]
As we don't have any present plans to make these cache types accessible via OKAPI, I would go with implementing just the one you need. There's nothing wrong with implementing `status3` later on, if it proves necessary.

Probably, most of the clients will keep using the original (deprecated) `status` either way.

**Comment by wrygiel** [comment]
Although, we might wonder if adding "Other" as a possible value to `status2` might be useful in terms of basic future-compatibility?

**Comment by wrygiel** [comment]
We might also **require** the clients to expect **any** value in `status2` (similar to the `type` field).

**Comment by following5** [comment]
> As we don't have any present plans to make these cache types accessible via OKAPI, I would go with implementing just the one you need. 

Currently I don't need any of them. Noone (except for you, posting #4 ;) has requested to make locked caches availble via OKAPI. I just expect that someone will need 'locked/visible' in the future, and maybe also the other status.

The only request so far is to return a better message than "This cache does not exist" if someone explicitly requests a locked cache. This also applies to any other unsupported status, so if we don't implement all of them, the error message should be changed anyway.

> We might also require the clients to expect any value in status2 (similar to the type field).

It's formally similar, but semantically different. There is a cache type "other" - also called "unknown" - on all OC nodes exactly for the purpose to represent unknown types. OKAPI itself maps unknown types to 'Unknown' (which should be documented; cmp #220). For sizes, there is a similar option called "other".

But there is nothing like an "unknown status", and nothing else an unknown status could be naturally mapped to. E.g. Opencaching.de is considering a new status like "Maybe available"*, which means: There is a near-50:50 chance that the cache is gone. It is in the middle between "Available" and "Temporarily unavailable", and the the fallback can be one of those two. OKAPI currently would map it to "Archived" (core.php line 1733)  and thus exclude the caches from many searches which should include it. This would be a bug.

Do you undestand this? **Mapping an unknown status to a known status always breaks things.** Therefore we should not require that developers do that, and we should avoid it within OKAPI.

**Comment by wrygiel** [comment]
> Currently I don't need any of them. Noone (except for you, posting #4 ;) has requested to make locked caches availble via OKAPI.

Wait, I'm a little lost. I was thinking that you wanted it too ;)

I think that we need an online chat (or audio) session. I'm starting to be afraid that we're wasting time here, and planning to implement something that none of us really wants implemented ;)

Will today 22:00CET on Skype suit you? My login: wrygiel3

**Comment by wrygiel** [comment]
Or, even earlier, 21:30CET? I forgot I had a Skype meeting with some friends planned for 22:15 :)

**Comment by following5** [comment]
I want it, but it's low prio. Im am currently too fucked up for any other kind of conversation, but we can postpone this issue.

**Comment by following5** [comment]
An interesting detail:

The old OC.de API - the "XML interface" - allows access to any hidden caches except caches which has not yet been published. But it will return "" or 0 for the following fields:

- cache title
- cache description, short description and hint
- cache coordinates
- image titles and urls

**Comment by following5** [comment]
The change that Bananeweizen requested was implemented. We may open a new issue for querying locked caches via OKAPI.
